### DISCUSSION TITLE
shadowhook_hook_sym_name Hook 成功但是走不到代理函数里面

### DISCUSSION THREAD

**Discussion by 16480000** [bytedance/android-inline-hook#37] [Q&A]
Hook 日志：
shadowhook_tag: shadowhook: hook_sym_name(libXXX.so, symName, 0x79da8e09b0) ...
shadowhook_tag: shadowhook: hook_sym_name(libXXX.so, symName, 0x79da8e09b0) OK. return: 0x7a5611b180. 1 - Pending task
HOOK-log: inline_hook success func symName 
被Hook的符号使用readelf -a 无法获取，只能通过objdump -x 来获取，是否和符号被隐藏有关

**Comment by 16480000** [comment]
是不是未导出的符号不能处理
shadowhook_tag: shadowhook: hook_sym_name(libXXX.so, symName, 0x79e4821b28) FAILED. 18 - Find symbol in ELF failed

**Comment by caikelun** [comment]
`FAILED. 18 - Find symbol in ELF failed` 说明这个 so 库中没有这个符号。

可以通过下面的命令查看 so 中有哪些符号，也就是可以用 `shadowhook_hook_sym_name` 来 hook 的函数：

1. `.dynsym` 和 `.symtab` 中的符号

```
llvm-readelf -sW ./libXXX.so | grep -v " UND "
```

2. `.gnu_debugdata` 中的符号

先确认下有没有 `.gnu_debugdata` section：
```
llvm-readelf -S ./libXXX.so | grep ".gnu_debugdata"
```

有的话：
```
llvm-objcopy --dump-section .gnu_debugdata=libXXX_gnu_debugdata.so.xz ./libXXX.so
xz -d -k libXXX_gnu_debugdata.so.xz
llvm-readelf -sW ./libXXX_gnu_debugdata.so | grep -v " UND "
```

另外，如果函数在 so 中没有符号，只要能在运行时获取到函数的绝对地址，也可以用 `shadowhook_hook_func_addr` 来 hook。

**Reply by 16480000** [reply]
找到原因了，是符号未导出导致无法hook，再请教一个问题，hook下来之后为什么无法在调用原函数之前执行调用函数或者给全局变量赋值的操作
